File: scripts/sync-finance-rs/src/pds.rs
//! ATProto PDS client — plain HTTP, no SDK.

use std::cell::RefCell;

use anyhow::{bail, Context, Result};
use serde_json::Value;

const PUBLIC_API: &str = "https://public.api.bsky.app";

pub struct PdsClient {
    pub did: String,
    pds: String,
    session: RefCell<Session>,
}

struct Session {
    access_jwt: String,
    refresh_jwt: String,
}

impl PdsClient {
    pub fn new(
        client: &reqwest::blocking::Client,
        handle: &str,
        app_password: &str,
    ) -> Result<Self> {
        // Resolve handle -> DID
        let resp: Value = client
            .get(format!(
                "{PUBLIC_API}/xrpc/com.atproto.identity.resolveHandle"
            ))
            .query(&[("handle", handle)])
            .send()?
            .error_for_status()?
            .json()?;
        let did = resp["did"]
            .as_str()
            .context("missing did in resolveHandle response")?
            .to_string();

        // Resolve DID -> PDS endpoint
        let doc: Value = if did.starts_with("did:plc:") {
            client
                .get(format!("https://plc.directory/{did}"))
                .send()?
                .error_for_status()?
                .json()?
        } else if did.starts_with("did:web:") {
            let host = did.strip_prefix("did:web:").unwrap().replace(':', "/");
            client
                .get(format!("https://{host}/.well-known/did.json"))
                .send()?
                .error_for_status()?
                .json()?
        } else {
            bail!("Unsupported DID method: {did}");
        };

        let pds = doc["service"]
            .as_array()
            .context("missing service in DID doc")?
            .iter()
            .find(|s| s["type"].as_str() == Some("AtprotoPersonalDataServer"))
            .context("no AtprotoPersonalDataServer in DID doc")?["serviceEndpoint"]
            .as_str()
            .context("missing serviceEndpoint")?
            .to_string();

        // Create session
        let sess: Value = client
            .post(format!("{pds}/xrpc/com.atproto.server.createSession"))
            .json(&serde_json::json!({
                "identifier": handle,
                "password": app_password,
            }))
            .send()?
            .error_for_status()?
            .json()?;

        let access_jwt = sess["accessJwt"]
            .as_str()
            .context("missing accessJwt")?
            .to_string();
        let refresh_jwt = sess["refreshJwt"]
            .as_str()
            .context("missing refreshJwt")?
            .to_string();

        println!("  Authenticated as {handle} ({did})");

        Ok(Self {
            did,
            pds,
            session: RefCell::new(Session {
                access_jwt,
                refresh_jwt,
            }),
        })
    }

    pub fn put_record(
        &self,
        client: &reqwest::blocking::Client,
        collection: &str,
        rkey: &str,
        record: &Value,
    ) -> Result<()> {
        let body = serde_json::json!({
            "repo": self.did,
            "collection": collection,
            "rkey": rkey,
            "record": record,
            "validate": false,
        });

        let resp = client
            .post(format!(
                "{}/xrpc/com.atproto.repo.putRecord",
                self.pds
            ))
            .header(
                "Authorization",
                format!("Bearer {}", self.session.borrow().access_jwt),
            )
            .json(&body)
            .send()?;

        if resp.status().as_u16() == 400 {
            let text = resp.text()?;
            if text.to_lowercase().contains("expired") {
                self.refresh(client)?;
                return self.put_record(client, collection, rkey, record);
            }
            bail!("PDS putRecord 400: {text}");
        }

        if !resp.status().is_success() {
            let status = resp.status();
            let text = resp.text().unwrap_or_default();
            bail!("PDS putRecord {status}: {text}");
        }

        Ok(())
    }

    fn refresh(&self, client: &reqwest::blocking::Client) -> Result<()> {
        let resp: Value = client
            .post(format!(
                "{}/xrpc/com.atproto.server.refreshSession",
                self.pds
            ))
            .header(
                "Authorization",
                format!("Bearer {}", self.session.borrow().refresh_jwt),
            )
            .send()?
            .error_for_status()?
            .json()?;

        let mut session = self.session.borrow_mut();
        session.access_jwt = resp["accessJwt"]
            .as_str()
            .context("missing accessJwt in refresh")?
            .to_string();
        session.refresh_jwt = resp["refreshJwt"]
            .as_str()
            .context("missing refreshJwt in refresh")?
            .to_string();

        println!("  Session refreshed");
        Ok(())
    }
}

File: packages/atproto/pds.js
/**
 * Shared ATProto PDS client — resolves identities, manages sessions,
 * reads/writes records. No dependencies, pure fetch + WebCrypto.
 *
 * Usage (browser or Worker):
 *   import { resolveHandle, resolvePds, PdsClient } from '../../packages/atproto/pds.js';
 *
 *   const did = await resolveHandle('alice.bsky.social');
 *   const pds = await resolvePds(did);
 *   const client = new PdsClient(pds);
 *   await client.login('alice.bsky.social', 'app-password');
 *   await client.putRecord('com.example.thing', 'self', { hello: 'world' });
 */

const PUBLIC_API = 'https://public.api.bsky.app';
const PLC_DIRECTORY = 'https://plc.directory';

// ─── Identity Resolution ─────────────────────────────────────────

/** Resolve a Bluesky handle to a DID. Uses public API, no auth needed. */
export async function resolveHandle(handle) {
  handle = handle.replace(/^@/, '').trim();
  const res = await fetch(
    `${PUBLIC_API}/xrpc/com.atproto.identity.resolveHandle?handle=${encodeURIComponent(handle)}`
  );
  if (!res.ok) throw new Error(`Could not resolve handle: @${handle}`);
  const { did } = await res.json();
  return did;
}

/** Resolve a DID to a PDS service endpoint. Supports did:plc and did:web. */
export async function resolvePds(did) {
  let doc;
  if (did.startsWith('did:plc:')) {
    const res = await fetch(`${PLC_DIRECTORY}/${did}`);
    if (!res.ok) throw new Error(`Could not resolve DID: ${did}`);
    doc = await res.json();
  } else if (did.startsWith('did:web:')) {
    const domain = did.replace('did:web:', '');
    const res = await fetch(`https://${domain}/.well-known/did.json`);
    if (!res.ok) throw new Error(`Could not resolve DID: ${did}`);
    doc = await res.json();
  } else {
    throw new Error(`Unsupported DID method: ${did}`);
  }

  const svc = doc.service?.find((s) => s.id === '#atproto_pds');
  if (!svc) throw new Error(`No PDS found for ${did}`);
  return svc.serviceEndpoint;
}

// ─── TID Generation ──────────────────────────────────────────────

/** Generate an ATProto TID (timestamp-based record key). */
export function generateTid() {
  const now = BigInt(Date.now()) * 1000n; // microseconds
  const clockId = BigInt(Math.floor(Math.random() * 1024)); // 10 random bits
  const tid = (now << 10n) | clockId;
  return tid.toString(36).padStart(13, '0');
}

// ─── PDS Client ──────────────────────────────────────────────────

export class PdsClient {
  /** @param {string} service - PDS URL (e.g. https://bsky.social) */
  constructor(service) {
    this.service = service.replace(/\/$/, '');
    if (!this.service.startsWith('http')) {
      this.service = `https://${this.service}`;
    }
    /** @type {{ did: string, handle: string, accessJwt: string, refreshJwt: string } | null} */
    this.session = null;
  }

  /** Authenticate with handle + app password. */
  async login(handle, appPassword) {
    const res = await this._xrpc('com.atproto.server.createSession', {
      identifier: handle,
      password: appPassword,
    });
    this.session = res;
    return this.session;
  }

  /** Refresh an expired access token. */
  async refreshSession() {
    if (!this.session) throw new Error('No session to refresh');
    const res = await fetch(
      `${this.service}/xrpc/com.atproto.server.refreshSession`,
      {
        method: 'POST',
        headers: { Authorization: `Bearer ${this.session.refreshJwt}` },
      }
    );
    if (!res.ok) throw new Error(`Refresh failed: ${res.status}`);
    const data = await res.json();
    this.session = { ...this.session, ...data };
  }

  /** Restore a previously stored session (for durable login). */
  restoreSession(session) {
    this.session = session;
  }

  /** Get the current session info. */
  getSession() {
    return this.session;
  }

  /** Get the service URL. */
  getService() {
    return this.service;
  }

  // ─── Record Operations ───────────────────────────────────────

  /** Get a single record. Returns null if not found. */
  async getRecord(collection, rkey) {
    if (!this.session) throw new Error('Not authenticated');
    const params = new URLSearchParams({
      repo: this.session.did,
      collection,
      rkey,
    });
    const res = await fetch(
      `${this.service}/xrpc/com.atproto.repo.getRecord?${params}`,
      { headers: { Authorization: `Bearer ${this.session.accessJwt}` } }
    );
    if (res.status === 404 || res.status === 400) return null;
    if (!res.ok) throw new Error(`getRecord failed: ${res.status}`);
    return res.json();
  }

  /** Get a record from a specific repo (can be another user's DID). */
  async getRecordFrom(did, collection, rkey) {
    const params = new URLSearchParams({ repo: did, collection, rkey });
    const headers = {};
    if (this.session) headers['Authorization'] = `Bearer ${this.session.accessJwt}`;
    const res = await fetch(
      `${this.service}/xrpc/com.atproto.repo.getRecord?${params}`,
      { headers }
    );
    if (res.status === 404 || res.status === 400) return null;
    if (!res.ok) throw new Error(`getRecord failed: ${res.status}`);
    return res.json();
  }

  /** List records in a collection. */
  async listRecords(collection, limit = 100, cursor) {
    if (!this.session) throw new Error('Not authenticated');
    const params = new URLSearchParams({
      repo: this.session.did,
      collection,
      limit: String(limit),
    });
    if (cursor) params.set('cursor', cursor);
    const res = await fetch(
      `${this.service}/xrpc/com.atproto.repo.listRecords?${params}`,
      { headers: { Authorization: `Bearer ${this.session.accessJwt}` } }
    );
    if (!res.ok) throw new Error(`listRecords failed: ${res.status}`);
    return res.json();
  }

  /** List records from a specific repo (can be another user's DID). */
  async listRecordsFrom(did, collection, limit = 100, cursor) {
    const params = new URLSearchParams({
      repo: did,
      collection,
      limit: String(limit),
    });
    if (cursor) params.set('cursor', cursor);
    const headers = {};
    if (this.session) headers['Authorization'] = `Bearer ${this.session.accessJwt}`;
    const res = await fetch(
      `${this.service}/xrpc/com.atproto.repo.listRecords?${params}`,
      { headers }
    );
    if (!res.ok) throw new Error(`listRecords failed: ${res.status}`);
    return res.json();
  }

  /** Create or update a record at a specific rkey. */
  async putRecord(collection, rkey, record) {
    return this._xrpc('com.atproto.repo.putRecord', {
      repo: this.session.did,
      collection,
      rkey,
      record,
    });
  }

  /** Create a record with an auto-generated rkey (TID). */
  async createRecord(collection, record) {
    return this._xrpc('com.atproto.repo.createRecord', {
      repo: this.session.did,
      collection,
      record,
    });
  }

  /** Delete a record. */
  async deleteRecord(collection, rkey) {
    await this._xrpc('com.atproto.repo.deleteRecord', {
      repo: this.session.did,
      collection,
      rkey,
    });
  }

  // ─── Blob Operations ─────────────────────────────────────────

  /** Upload a blob to the PDS. Returns the blob ref for embedding in records. */
  async uploadBlob(data, mimeType) {
    if (!this.session) throw new Error('Not authenticated');
    const res = await fetch(
      `${this.service}/xrpc/com.atproto.repo.uploadBlob`,
      {
        method: 'POST',
        headers: {
          Authorization: `Bearer ${this.session.accessJwt}`,
          'Content-Type': mimeType,
        },
        body: data instanceof ArrayBuffer ? data : data.buffer,
      }
    );
    if (!res.ok) {
      const text = await res.text();
      throw new Error(`uploadBlob failed (${res.status}): ${text}`);
    }
    const { blob } = await res.json();
    return blob;
  }

  /** Fetch a blob by DID + CID. Works cross-PDS for public blobs. */
  async getBlob(did, cid) {
    const params = new URLSearchParams({ did, cid });
    const headers = {};
    if (this.session) headers['Authorization'] = `Bearer ${this.session.accessJwt}`;
    const res = await fetch(
      `${this.service}/xrpc/com.atproto.sync.getBlob?${params}`,
      { headers }
    );
    if (!res.ok) throw new Error(`getBlob failed (${res.status})`);
    const buf = await res.arrayBuffer();
    return new Uint8Array(buf);
  }

  // ─── Internal ─────────────────────────────────────────────────

  /** Generic XRPC POST call with auth and structured error handling. */
  async _xrpc(method, body) {
    if (!this.session && !method.includes('createSession')) {
      throw new Error('Not authenticated');
    }
    const headers = { 'Content-Type': 'application/json' };
    if (this.session) {
      headers['Authorization'] = `Bearer ${this.session.accessJwt}`;
    }
    const res = await fetch(`${this.service}/xrpc/${method}`, {
      method: 'POST',
      headers,
      body: JSON.stringify(body),
    });
    if (!res.ok) {
      const text = await res.text();
      let detail = '';
      try {
        const err = JSON.parse(text);
        const code = err.error || '';
        const msg = err.message || '';
        if (code === 'AuthFactorTokenRequired') {
          detail = 'Two-factor auth is enabled. Use an app password instead.';
        } else if (code === 'RateLimitExceeded') {
          detail = 'Rate limited. Wait a moment and try again.';
        } else if (msg.includes('Invalid identifier or password')) {
          detail = 'Invalid handle or app password.';
        } else {
          detail = code ? `${code}: ${msg}` : msg || text;
        }
      } catch {
        detail = text || `HTTP ${res.status}`;
      }
      throw new Error(detail || `XRPC ${method} failed (${res.status})`);
    }
    return res.json();
  }
}
